File: veilclient/test/integration_tests/sample_test.cc
/**
 * @file sample_test.cc
 * @author Rafal Slota
 * @copyright (C) 2013 ACK CYFRONET AGH
 * @copyright This software is released under the MIT license cited in 'LICENSE.txt'
 */

#include "erlTestCore.h"
#include "fslogicProxy.h"
#include "oneErrors.h"
#include "testCommon.h"

#include <boost/filesystem.hpp>

using namespace boost::filesystem;

// TEST definitions below

class SampleTest: public CommonIntegrationTest
{
protected:
    path directIO_root;

    // Mount file system in "main" subdir with "peer.pem" cert
    // use onedataMount->getRoot() to get absolute mount point
    SampleTest()
        : CommonIntegrationTest{std::unique_ptr<one::testing::FsImplMount>{new one::testing::FsImplMount{"main", "peer.pem"}}}
    {
    }

    void SetUp() override
    {
        // Initialization of the whole client.
        // This initialization is not required if test uses only filesystem (theres no oneclient code calls)
        CommonIntegrationTest::SetUp();
    }
};

// This test shows how you can call sample_test:exec/1 method on cluster environment
TEST_F(SampleTest, clusterCommandExec) {
    EXPECT_EQ(std::string("/tmp/dio"), one::testing::erlExec("{env, \"DIO_ROOT\"}"));
}

// onedataMount->getRoot() is set to the root of mounted FsImpl. Therefore you can just
// manage some files in order to test whole oneclient behaviourally
TEST_F(SampleTest, mkdirExample) {
    EXPECT_EQ(0, ::system(("mkdir " + onedataMount->getRoot() + "/testDir").c_str()));
    EXPECT_EQ(0, ::system(("rm -rf " + onedataMount->getRoot() + "/testDir").c_str()));
}

TEST_F(SampleTest, fileExample) {
    EXPECT_EQ(0, ::system(("touch " + onedataMount->getRoot() + "/file").c_str()));
}

TEST_F(SampleTest, fslogicExample) {
    EXPECT_EQ(VOK, fslogic->createDir("testDir", 0755));
}

File: veilclient/test/integration_tests/sample_test.erl
%% ===================================================================
%% @author Rafal Slota
%% @copyright (C): 2013 ACK CYFRONET AGH
%% This software is released under the MIT license 
%% cited in 'LICENSE.txt'.
%% @end
%% ===================================================================
%% @doc: SampleTest cluster side test driver.      
%% @end
%% ===================================================================
-module(sample_test).
-include("test_common.hrl").

-export([setup/1, teardown/2, exec/1]).


%% This method will be called before each test suite (suite  = whole test translation unit, one .cc test file - one setup/1 call)
setup(ccm) ->
    ok;
setup(worker) ->
    DirectIORoot = "/tmp/dio",
    os:putenv("DIO_ROOT", DirectIORoot), 
    Fuse_groups = [{fuse_group_info, "cluster_fid", {storage_helper_info, "DirectIO", [DirectIORoot]}}],
    fslogic_storage:insert_storage("ClusterProxy", [], Fuse_groups),
    
    test_common:register_user("peer.pem"),

    DirectIORoot.


%% Same as setup/1 except that teardown is called after test suite
%% State argument is an return value of setup/1
teardown(ccm, _State) ->
    ok; 
teardown(worker, _State) ->
    ok.


%% This method will be called on a cluster when you use erlExec(string) function in test suite
%% C++ erlEcec string argument will be translated to erlang term (e.g. "{1, ok}" will eval to {1, ok}). 
%% Also you have to return a string. All other terms will be translated to string using io:format 
exec({env, VarName}) ->
    os:getenv(VarName).
